hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/merge/SparkRecordMergingUtils.java

```diff
@@ -0,0 +1,189 @@
+        Integer ordInPartialUpdate = newPartialNameToIdMapping.get(structField.name());
+        if (ordInPartialUpdate != null) {
+          // pick from new
+          values.add(newPartialRow.get(ordInPartialUpdate, structField.dataType()));
+        } else {
+          // pick from old
+          values.add(oldRow.get(fieldId, structField.dataType()));
+        }
+      }
+      InternalRow mergedRow = new GenericInternalRow(values.toArray());
+
+      HoodieSparkRecord mergedSparkRecord = new HoodieSparkRecord(
+          mergedRow, mergedSchemaPair.getRight().getLeft());
+      return Pair.of(mergedSparkRecord, mergedSchemaPair.getRight().getRight());
+    } else {
+      return Pair.of(newer, newSchema);
+    }
+  }
+
+  /**
+   * @param avroSchema Avro schema.
+   * @return The field ID to {@link StructField} instance mapping.
+   */
+  public static Map<Integer, StructField> getCachedFieldIdToFieldMapping(Schema avroSchema) {
+    return FIELD_ID_TO_FIELD_MAPPING_CACHE.computeIfAbsent(avroSchema, schema -> {
+      StructType structType = HoodieInternalRowUtils.getCachedSchema(schema);
+      Map<Integer, StructField> schemaFieldIdMapping = new HashMap<>();
+      int fieldId = 0;
+
+      for (StructField field : structType.fields()) {
+        schemaFieldIdMapping.put(fieldId, field);
+        fieldId++;
+      }
+
+      return schemaFieldIdMapping;
+    });
+  }
+
+  /**
+   * @param avroSchema Avro schema.
+   * @return The field name to ID mapping.
+   */
+  public static Map<String, Integer> getCachedFieldNameToIdMapping(Schema avroSchema) {
+    return FIELD_NAME_TO_ID_MAPPING_CACHE.computeIfAbsent(avroSchema, schema -> {
+      StructType structType = HoodieInternalRowUtils.getCachedSchema(schema);
+      Map<String, Integer> schemaFieldIdMapping = new HashMap<>();
+      int fieldId = 0;
+
+      for (StructField field : structType.fields()) {
+        schemaFieldIdMapping.put(field.name(), fieldId);
+        fieldId++;
+      }
+
+      return schemaFieldIdMapping;
+    });
+  }
+
+  /**
+   * Merges the two schemas so the merged schema contains all the fields from the two schemas,
+   * with the same ordering of fields, and the fields of the old schema comes first.
+   *
+   * @param oldSchema Old schema.
+   * @param newSchema New schema.
+   * @return The ID to {@link StructField} instance mapping of the merged schema, and the
+   * {@link StructType} and Avro schema of the merged schema.
+   */
+  public static Pair<Map<Integer, StructField>, Pair<StructType, Schema>> getCachedMergedSchema(Schema oldSchema,
+                                                                                                Schema newSchema,
+                                                                                                Schema readerSchema) {
+    return MERGED_SCHEMA_CACHE.computeIfAbsent(
+        Pair.of(Pair.of(oldSchema, newSchema), readerSchema), schemaPair -> {
+          Schema schema1 = schemaPair.getLeft().getLeft();
```

Comment:
Replace pair(pair, single) with Tuple3 ? Hmm, our Tuple3 does not implement the `#equals` yet.

Reply:
Right, I prefer not to change that in this PR to avoid any unintended changes.  So I use Java `Pair` class instead.